### PR TITLE
Add combo hook to allow per layer combo reference layers.

### PR DESCRIPTION
**Pull request by EricGebhart**
This adds a hook function so that a combo reference layer can be set
for the current layer.  It adds a weak function to process_combos.c, 
and calls that function in process_combo the same way the ref_only
define is managed.  Also adds some macro code to gboards/g/keymap_combos to enable the function to be defined with 
the def file.  docs were extended.

## Description
Add a weak function to process_combo which takes a layer and
returns the layer.  If a default layer is defined call it with that, otherwise
call it with the current layer.  If the layer is not the same as the current
layer replace the keycode with the one from the reference.  This is
the same mechanism as the COMBO_ONLY_FROM_LAYER.
if COMBO_ONLY_FROM_LAYER is set, this new code is skipped.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: docs/feature_combo.md [hunk 1]
```diff
@@ -328,6 +328,36 @@ If you, for example, use multiple base layers for different key layouts, one for
...
+will give a reference layer of _QWERTY for the _DVORAK layer and
+will give the _NAV layer as a reference to it's self. All other layers
+will have the default for their combo reference layer. If the default
+is not set, all other layers will reference themselves.
+
+    ```c
+    #define COMBO_REF_DEFAULT _MY_COMBO_REF_LAYER
+    ...
+
+    uint16_t combo_ref_from_layer(uint16_t layer){
```
Comment (drashna) [inline docs/feature_combo.md:351]: QMK only supports up to 16 layers, so shouldn't uint8_t be plenty here? 
```suggestion
    uint16_t combo_ref_from_layer(uint8_t layer){
```

File: docs/feature_combo.md [hunk 2]
```diff
@@ -328,6 +328,36 @@ If you, for example, use multiple base layers for different key layouts, one for
...
+    switch (biton32(layer_state)){
```
Comment (drashna) [inline docs/feature_combo.md:352]:
```suggestion
    switch (get_highest_layer(layer_state)){
```

File: keyboards/gboards/g/keymap_combo.h [hunk 3]
```diff
@@ -66,10 +69,30 @@ void process_combo_event(uint16_t combo_index, bool pressed) {
...
+// Allow reference layers per layer.
+#define COMB BLANK
+#define SUBS BLANK
+#define TOGG BLANK
+
+#undef COMBO_REF_LAYER
+#define COMBO_REF_LAYER REF_LAYER
+
+uint16_t combo_ref_from_layer(uint16_t layer){
+  switch (biton32(layer_state)){
```
Comment (drashna) [inline keyboards/gboards/g/keymap_combo.h:89]:
```suggestion
uint16_t combo_ref_from_layer(uint8_t layer){
  switch (get_highest_layer(layer_state)){
```

File: quantum/process_keycode/process_combo.c [hunk 4]
```diff
@@ -547,6 +554,16 @@ bool process_combo(uint16_t keycode, keyrecord_t *record) {
 #ifdef COMBO_ONLY_FROM_LAYER
     /* Only check keycodes from one layer. */
     keycode = keymap_key_to_keycode(COMBO_ONLY_FROM_LAYER, record->event.key);
+#else
+
+#    ifdef COMBO_REF_DEFAULT
+    layer          = combo_ref_from_layer(COMBO_REF_DEFAULT);
+#    else
+    layer = combo_ref_from_layer(biton32(layer_state));
```
Comment (drashna) [inline quantum/process_keycode/process_combo.c:562]:
```suggestion
    layer = combo_ref_from_layer(get_highest_layer(layer_state));
```

File: quantum/process_keycode/process_combo.c [hunk 5]
```diff
@@ -547,6 +554,16 @@ bool process_combo(uint16_t keycode, keyrecord_t *record) {
...
+#    endif
+    if (layer != biton32(layer_state)) {
```
Comment (drashna) [inline quantum/process_keycode/process_combo.c:564]:
```suggestion
    if (layer != get_highest_layer(layer_state)) {
```

File: quantum/process_keycode/process_combo.c [hunk 6]
```diff
@@ -29,6 +29,12 @@ extern uint16_t COMBO_LEN;
 
 __attribute__((weak)) void process_combo_event(uint16_t combo_index, bool pressed) {}
 
+#ifndef COMBO_ONLY_FROM_LAYER
+__attribute__((weak)) uint16_t combo_ref_from_layer(uint16_t layer) {
```
Comment (drashna) [inline quantum/process_keycode/process_combo.c:33]:
```suggestion
__attribute__((weak)) uint16_t combo_ref_from_layer(uint8_t layer) {
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

File: quantum/process_keycode/process_combo.c [hunk 7]
```diff
@@ -547,6 +554,16 @@ bool process_combo(uint16_t keycode, keyrecord_t *record) {
...
+#else
+
+#    ifdef COMBO_REF_DEFAULT
+    layer          = combo_ref_from_layer(COMBO_REF_DEFAULT);
+#    else
+    layer = combo_ref_from_layer(get_highest_layer(layer_state));
+#    endif
+    if (layer != get_highest_layer(layer_state)) {
+        keycode = keymap_key_to_keycode(layer, record->event.key);
+    }
```
Comment (sigprof) [inline quantum/process_keycode/process_combo.c:566]:
```suggestion
    uint8_t highest_layer = get_highest_layer(layer_state);
#    ifdef COMBO_REF_DEFAULT
    layer = combo_ref_from_layer(COMBO_REF_DEFAULT);
#    else
    layer = combo_ref_from_layer(highest_layer);
#    endif
    if (layer != highest_layer) {
        keycode = keymap_key_to_keycode(layer, record->event.key);
    }
```
With this kind of code there is some chance that the compiler would be able to optimize the unused code away if LTO is enabled (without LTO the optimizer won't be able to assume anything about the implementation of a weak function).

**Review by sigprof — CHANGES_REQUESTED** [review]
The feature itself looks like it could be useful in some cases; however, there are some questionable implementation choices.

The `COMBO_REF_DEFAULT` feature duplicates `COMBO_ONLY_FROM_LAYER` if a custom `combo_ref_from_layer()` function is not actually defined (with a subtle difference due to the `if (layer != get_highest_layer(layer_state))` test, but that's mostly an implementation detail, and any behavior differences might be possible only if the specified layer contains some `KC_TRNS` keycodes, which should not be the case for a reference layer).

However, if a custom `combo_ref_from_layer()` function is defined, the `COMBO_REF_DEFAULT` feature is not actually needed, because the user could achieve the same result by adding `default: return _MY_REF_LAYER;` to that function instead (and that would actually work better, because a layer enum constant may be used in that function, while `COMBO_REF_DEFAULT` would require a raw layer number).  The code for `combos.def` would need a new macro to specify the default reference layer though.

If we choose to remove `COMBO_REF_DEFAULT`, the possible API options are:
1. Always pass `get_highest_layer(layer_state)` to `combo_ref_from_layer()`, remove `get_highest_layer(layer_state)` from the example implementation of `combo_ref_from_layer()` (because the `layer` argument could be used instead).  This makes the user code shorter, but still leaves special handling for the current layer number.
2. Remove the argument from `combo_ref_from_layer()`, treat the return value of `MAX_LAYER` (which is always an invalid layer number) as a request to keep the current keycode.  The problem here is that such usage of `MAX_LAYER` might not be obvious.

Another concern is that the code size in the default case (no extra options specified) would increase slightly, because the compiler probably won't be able to optimize away the added code.  But I'm not sure whether this is important enough to add a `#define` for every new option.

**Comment by EricGebhart** [comment]
Thank you SigProf!

I was hoping to have this conversation with someone...

I introduced  `COMBO_REF_DEFAULT`  to highlight the ambiguity of  `COMBO_ONLY_FROM_LAYER` which is misleading
in the context of `combo_ref_from_layer()` where it becomes the default, not the __ONLY__.
I agree that it is unnecessary and I like the idea of letting the function do that work.  Which comes back to the `COMBO_ONLY_FROM_LAYER` which I think should probably be deprecated as it is also a Pain to use as it also needs
an integer value.  Probably better to just let the callback function be the proper way to assign a default/only reference layer as needed.

I'm not opposed to using `MAX_LAYER` but maybe define it with a name that is more meaningful in this context.
That does seems a bit magical to me, and not as clear as just explicitly saying what you would like in the callback.

The other thing might be that the name `combo_ref_from_layer` might also need simplifying, as it is not necessarily
based on layer, although that is the most obvious use.

So remove `COMBO_REF_DEFAULT` and pass layer in to `combo_ref_from_layer`  and make a note about `COMBO_ONLY_FROM_LAYER` being unnecessary.  And give the function template example to accomplish the same thing.

Does that sound reasonable ?

**Comment by EricGebhart** [comment]
This looks better.  There is still the conflict in behavior when COMBO_ONLY_FROM_LAYER is set.  I added the example function to the doc which accomplishes the same thing.  There needs to be some default return from the automatically created hook function in the keymap_combos.h  So there is a way to specify that. It also defaults to what was passed in so even if the function is created and is completely empty it will still work.

I do wonder if the hook function name is a good one.  Just because we pass the layer in doesn't mean that the decision has to be based on that.  Maybe a more general name would be better.

**Comment by EricGebhart** [comment]
If there is any question, I have been using this code since sometime in January.   I initially had two locales/languages with two combo reference layers, one for each locale, while also using some layers as their own reference. I have also turned off the reference layers in some cases.  pro-micro memory limits...  It has been very nice that I can change my keymaps or my locale as I wish without impacting my combos.  It also allows my navigation layer to have a different set of combos than all the rest of my layers.

I just tested it in various different ways, no reference, a reference, a layer which references its self. I am using the gboard macros which create the hook function automatically. 

The CI seems happy.   Let me know what you think of this version.  I think it is better.  It does bother me a little that `COMBO_ONLY_FROM_LAYER` is counter indicated, and a pain to use, but I think it's ok since the interaction is well documented.   It could be used as the default over the current layer, but the name of it is misleading for that purpose.

**Comment by EricGebhart** [comment]
that is a strange error in rgb_matrix.h  can't see how it's related to this.

**Comment by fauxpark** [comment]
It's not; the CI is trying to build all default and VIA keymaps (because you have modified core code), and that particular board is broken due to `DRIVER_LED_COUNT` being renamed to `RGB_MATRIX_LED_COUNT` in develop but not master. New boards added since that rename need to be fixed up in develop, until the next breaking changes merge at the end of the month.

**Comment by EricGebhart** [comment]
Oh, I think I merged from the wrong branch.

**Comment by EricGebhart** [comment]
Ok should be good to go again. Reviewed changes.

**Comment by EricGebhart** [comment]
Trying to figure out why it shows changes requested when I already incorporated them.

File: quantum/process_keycode/process_combo.c [hunk 8]
```diff
@@ -549,6 +556,13 @@ bool process_combo(uint16_t keycode, keyrecord_t *record) {
 #ifdef COMBO_ONLY_FROM_LAYER
     /* Only check keycodes from one layer. */
     keycode = keymap_key_to_keycode(COMBO_ONLY_FROM_LAYER, record->event.key);
+#else
+
+    uint8_t highest_layer = get_highest_layer(layer_state);
+    layer                 = combo_ref_from_layer(highest_layer);
+    if (layer != highest_layer) {
+        keycode = keymap_key_to_keycode(layer, record->event.key);
+    }
```
Comment (sigprof) [inline quantum/process_keycode/process_combo.c:565]: Looks like the `layer` variable is used only by this part of code, so you may as well declare it here (and maybe rename to `ref_layer` to be more descriptive):
```suggestion
    uint8_t highest_layer = get_highest_layer(layer_state);
    uint8_t ref_layer     = combo_ref_from_layer(highest_layer);
    if (ref_layer != highest_layer) {
        keycode = keymap_key_to_keycode(ref_layer, record->event.key);
    }
```

File: quantum/process_keycode/process_combo.c [hunk 9]
```diff
@@ -528,8 +534,9 @@ static bool process_single_combo(combo_t *combo, uint16_t keycode, keyrecord_t *
 }
 
 bool process_combo(uint16_t keycode, keyrecord_t *record) {
-    bool is_combo_key          = false;
-    bool no_combo_keys_pressed = true;
+    uint8_t layer;
+    bool    is_combo_key          = false;
+    bool    no_combo_keys_pressed = true;
```
Comment (sigprof) [inline quantum/process_keycode/process_combo.c:539]: If the change to declare `ref_layer` is done, this block can then be reverted to its original state.
```suggestion
    bool is_combo_key          = false;
    bool no_combo_keys_pressed = true;
```
Also in the `COMBO_ONLY_FROM_LAYER` case the compiler would probably complain about the unused `layer` variable declared here.

File: keyboards/gboards/g/keymap_combo.h [hunk 10]
```diff
@@ -66,10 +69,35 @@ void process_combo_event(uint16_t combo_index, bool pressed) {
...
+// Allow reference layers per layer.
+#define COMB BLANK
+#define SUBS BLANK
+#define TOGG BLANK
+
+#undef COMBO_REF_LAYER
+#define COMBO_REF_LAYER REF_LAYER
+
+uint8_t combo_ref_from_layer(uint8_t layer){
+  switch (get_highest_layer(layer_state)){
```
Comment (sigprof) [inline keyboards/gboards/g/keymap_combo.h:89]: Does this need to be `get_highest_layer(layer_state)` now (when the combo code no longer handles `COMBO_REF_DEFAULT`)?
```suggestion
  switch (layer) {
```

**Review by sigprof — CHANGES_REQUESTED** [review]
Apparently the documentation of `COMBO_REF_DEFAULT` does not match its actual usage after the latest modifications.

Another potential issue is that the `combos.def` support now unconditionally defines `combo_ref_from_layer()`, which some people might not want. Possible options here are:
1. Add some extra define to control the generation of `combo_ref_from_layer()` by `keymap_combo.h`.
2. Drop the `combo_ref_from_layer()` generation completely, requiring the users of `combos.def` to also define the function manually if they need it.
3. Change the default behavior of the generated `combo_ref_from_layer()` to be `return combo_ref_from_layer_custom(layer);` with yet another weak function.
4. Resort to some dirty hacks like `#define combo_ref_from_layer hidden_combo_ref_from_layer` to move the autogenerated function away, or `#define COMBO_REF_DEFAULT real_combo_ref_from_layer(layer)` to hook the default case instead of returning a constant.

A variation of case 1 that does not require a global define could be:
```c
#define COMBO_REF_LAYERS(defs)                    \
    uint8_t combo_ref_from_layer(uint8_t layer) { \
        switch (layer) { defs }                   \
        return layer;                             \
    }

#define REF_LAYER(from, to) \
    case from:              \
        return to;
#define DEFAULT_REF_LAYER(to) \
    default:                  \
        return to;
```
which would then be used like
```c
COMBO_REF_LAYERS(
    REF_LAYER(_DVORAK, _QWERTY)
    REF_LAYER(_NAV, _NAV)
    DEFAULT_REF_LAYER(_DEF)
)
```
(although `clang-format` won't like that, it probably won't get applied to `combos.def` anyway)

File: keyboards/gboards/g/keymap_combo.h [hunk 11]
```diff
@@ -66,10 +69,35 @@ void process_combo_event(uint16_t combo_index, bool pressed) {
...
 #undef SUBS
 #undef TOGG
+
+// Allow reference layers per layer.
+#define COMB BLANK
+#define SUBS BLANK
+#define TOGG BLANK
+
+#undef COMBO_REF_LAYER
+#define COMBO_REF_LAYER REF_LAYER
```
[the remark this answers is not in the document]
Comment (EricGebhart) [reply keyboards/gboards/g/keymap_combo.h:86]: oh, yea. good catch.  I actually have that in mine, I must not have updated last time.

File: docs/feature_combo.md [hunk 12]
```diff
@@ -328,6 +328,36 @@ If you, for example, use multiple base layers for different key layouts, one for
 
 With `#define COMBO_ONLY_FROM_LAYER 0` in config.h, the combos' keys are always checked from layer `0`, even if other layers are active.
 
+### Combo reference layers by layer.
+
+If not using `COMBO_ONLY_FROM_LAYER` it is possible to specify a
+combo reference layer for any layer using the `combo_ref_from_layer` hook. 
+This function receives the default reference layer if set, or the current
+layer otherwise. A default layer can be set with
+`#define COMBO_REF_DEFAULT _MY_COMBO_REF_LAYER`
```
[the remark this answers is not in the document]
Comment (EricGebhart) [reply docs/feature_combo.md:337]: yep. I need to remove COMBO_REF_DEFAULT the rest of the way.  I already have 
```
#define DEFAULT_REF_LAYER(LAYER)                \
  default: return LAYER;
```
I just didn't finish it correctly.

File: keyboards/gboards/g/keymap_combo.h (hunk 10 above)
Comment (EricGebhart) [reply keyboards/gboards/g/keymap_combo.h:89]: yep.  another thing I missed in my confusion of making it work for me, and making it work here. I didnt realize I had two versions.

Comment (EricGebhart) [reply keyboards/gboards/g/keymap_combo.h]: I already had this one..

File: keyboards/gboards/g/keymap_combo.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (EricGebhart) [reply keyboards/gboards/g/keymap_combo.h]: should be all good.

**Comment by EricGebhart** [comment]
This is looking pretty good. I'll come back and review again in a bit.

**Comment by EricGebhart** [comment]
Ok, reviewed. looks good.  

I might have been overly verbose in the doc.  

As far as the combo_ref_layer()  I think that if someone is using the macros they wont have a custom hook.  If we wanted more flexibility, I could add an combo_ref_inject.h  that could be pulled into the function in the same way the other combo function works.  That would allow for maximum flexibility, but it might be overkill.  But it would at least
be consistent with other combo macro behavior.

Take a look and let me know what you think.
thank you.

**Comment by EricGebhart** [comment]
I'm just looking at this to see if there is anything I need to do. There isn't as far as I can tell.
My last changes are over a month old and the current request for changes is completely out of date and no longer pertinent.

I would appreciate a re-review and some feedback. If we miss this development window this PR will be a year old.

**Review by tzarc — APPROVED** [review]

**Pull request merged**